enclave/api/src/lib.rs

```diff
@@ -0,0 +1,281 @@
+
+    //
+    // Grapevine-enclave specific
+    //
+
+    // CLIENT-FACING METHODS
+
+    /// Accept an inbound authentication request
+    fn client_accept(
+        &self,
+        req: ClientAuthRequest,
+    ) -> Result<(ClientAuthResponseWithChallengeSeed, ClientSession)>;
+
+    /// Destroy a peer association
+    fn client_close(&self, channel_id: ClientSession) -> Result<()>;
+
+    /// Service a user's encrypted QueryRequest
+    fn query(
+        &self,
+        payload: EnclaveMessage<ClientSession>,
+    ) -> Result<EnclaveMessage<ClientSession>>;
+
+    /// SERVER-FACING
+
+    /// Set the current timestamp, which is used to tag new records.
+    fn set_current_timestamp(&self, timestamp: u64) -> Result<()>;
+
+    /// Set the message time to live value, which is used to determine when a
+    /// record can be evicted.
+    fn set_message_time_to_live(&self, timestamp: u64) -> Result<()>;
+}
+
+/// Helper trait which reduces boiler-plate in untrusted side
+/// The trusted object which implements the above api usually cannot implement
+/// Clone, Send, Sync, etc., but the untrusted side can and usually having a
+/// "handle to an enclave" is what is most useful for a webserver.
+/// This marker trait can be implemented for the untrusted-side representation
+/// of the enclave.
+pub trait GrapevineEnclaveProxy: GrapevineEnclaveApi + Clone + Send + Sync + 'static {}
+
+impl<T> GrapevineEnclaveProxy for T where T: GrapevineEnclaveApi + Clone + Send + Sync + 'static {}
+
+// Error
+
+/// A generic result type for enclave calls
+pub type Result<T> = StdResult<T, Error>;
+
+/// An error when something goes wrong with handling a query
+#[derive(Serialize, Deserialize, Debug, Display, Clone)]
+pub enum QueryError {
+    /// Invalid message id length: {0}, expected {1}
+    InvalidMessageIdLength(usize, usize),
+    /// Invalid sender length: {0}
+    InvalidSenderLength(usize, usize),
+    /// Invalid recipient length: {0}
+    InvalidRecipientLength(usize, usize),
+    /// Invalid payload length: {0}
+    InvalidPayloadLength(usize, usize),
+    /// Invalid auth identity length: {0}, expected {1}
+    InvalidAuthIdentityLength(usize, usize),
+    /// Key error: {0}
+    Key(KeyError),
+    /// Signature conversion error
+    SignatureConversion,
+    /// Signature verification error
+    SignatureVerification,
+    /// Cannot update zeroes entry
+    CannotUpdateZeroesEntry,
+    /// Invalid request type
+    InvalidRequestType,
+}
+
+impl From<KeyError> for QueryError {
+    fn from(src: KeyError) -> Self {
+        Self::Key(src)
+    }
+}
+
+impl From<SignatureError> for QueryError {
+    fn from(_: SignatureError) -> Self {
+        Self::SignatureConversion
+    }
+}
+
+impl From<SchnorrkelError> for QueryError {
+    fn from(_: SchnorrkelError) -> Self {
+        Self::SignatureVerification
+    }
+}
+
+/// An error returned by the Grapevine enclave
+#[derive(Serialize, Deserialize, Debug, Display, Clone)]
+pub enum Error {
```

Comment:
Nit: No need for `error` everywhere, it is already implied from context. It also prevents stuff like `Foo error: Bar error: Frob error`